File: C++/66-机器人的运动范围.cpp
#include <vector>
using namespace std;
class Solution
{
public:
    int cnt;
    int rows;
    int cols;
    int movingCount(int threshold, int rows, int cols)
    {
        vector<bool> tmp(cols, false);
        vector<vector<bool> > arr(rows, tmp);
        cnt = 0;
        this->cols = cols;
        this->rows = rows;
        helper(arr, 0, 0, threshold);
        return cnt;
    }
    void helper(vector<vector<bool> > &visit, int i, int j, int key)
    {
        if (i < 0 || j < 0 || i >= rows || j >= cols || visit[i][j] == true)
            return;
        if (getSum(i) + getSum(j) > key)
            return;
        visit[i][j] = true;
        cnt++;
        helper(visit, i - 1, j, key);
        helper(visit, i + 1, j, key);
        helper(visit, i, j - 1, key);
        helper(visit, i, j + 1, key);
    }
    int getSum(int x)
    {
        int ret = 0;
        while (x)
        {
            ret += x % 10;
            x /= 10;
        }
        return ret;
    }
};

File: Python/66-机器人的运动范围.py
# -*- coding:utf-8 -*-
class Solution:
    def __init__(self):
        self.cnt = 0

    def movingCount(self, threshold, rows, cols):
        # write code here
        arr = [[1 for _ in range(cols)] for _ in range(rows)]
        self.helper(arr, 0, 0, threshold)
        return self.cnt

    def helper(self, visit, i, j, k):
        if i < 0 or j < 0 or i >= len(visit) or j >= len(visit[0]) or visit[i][j] == 0:
            return
        if sum(map(int, str(i))) + sum(map(int, str(j))) > k:
            return
        self.cnt += 1
        visit[i][j] = 0
        self.helper(visit, i - 1, j, k)
        self.helper(visit, i + 1, j, k)
        self.helper(visit, i, j - 1, k)
        self.helper(visit, i, j + 1, k)